src/opm/input/eclipse/Schedule/KeywordHandlers.cpp

```diff
@@ -972,7 +972,17 @@ File {} line {}.)", wname, location.keyword, location.filename, location.lineno)
             const std::string& wellNamePattern = record.getItem("WELL").getTrimmedString(0);
             const auto well_names = this->wellNames(wellNamePattern, handlerContext);
 
-            const Well::Status status = WellStatusFromString(record.getItem("STATUS").getTrimmedString(0));
+            Well::Status status = WellStatusFromString(record.getItem("STATUS").getTrimmedString(0));
+            // If status is OPEN, but we have only zero rates, we override it to be STOP instead.
+            // Note that comparing the UDAValues using '... == UDAValue(0.0)' will not work,
+            // since the rhs will have unity dimensions, and the ORAT etc. will not.
+            const bool allZeroRates =
```

Comment:
I did not read the UDA part, but from the code change here, it looks like we only mark this well to be `STOP` only when all the three rates are zeros. 

But in theory, if any rate constraint is zero, the well should be `STOP`. Is it right?

And furthermore, if WCONHIST is specified with `BHP` control, then the three rates are pure historic values, then well should be able to run.

Reply:
My current feeling is that we can only process to `STOP` wells with zero rate constraint only when we are sure what actual rate constraints are used (e.g. WHISCTL can change the control for all the historic producers), possibly after passing the whole `SCHEDULE` section.

Reply:
> But in theory, if any rate constraint is zero, the well should be `STOP`. Is it right?

I do not think that is necessarily true, if there is no water present in a region, a zero water rate constraint could make sense. In any case, the change here only deals with WCONHIST, so the data looked at are historical data, not constraints. It is common to see one of them being zero. For example from Norne:
```
     'D-1H'      'OPEN'      'RESV'   4347.700      0.000     482594.703  5* /
```

> And furthermore, if WCONHIST is specified with `BHP` control, then the three rates are pure historic values, then well should be able to run.

While possible, I do not think it makes sense to set BHP control for a well when you all zero historical rates, but if you really want to, you can do it with WCONPROD instead. I think it is better to set STOP for that well when using WCONHIST in that case (as done here).

> My current feeling is that we can only process to `STOP` wells with zero rate constraint only when we are sure what actual rate constraints are used (e.g. WHISCTL can change the control for all the historic producers), possibly after parsing the whole `SCHEDULE` section.

I disagree. I think we should not care at all about the control setting when all historical rates are zero, just set STOP always.

@vkip @tskille do you have an opinion?

Reply:
> I do not think that is necessarily true, if there is no water present in a region, a zero water rate constraint could make sense. 

Then you will get a singluar situation to solve.  The zero water constraint does not provide any constraint (only trouble) in theory. 

>  'D-1H'      'OPEN'      'RESV'   4347.700      0.000     482594.703  5* /

Yes. this zero is not constraint if the control is RESV, so this well should not be STOPPED. But if you change RESV to be WRAT here, then this well need to be STOPPED. It is the control type causing problems in the simulation, not the historical values. 

> While possible, I do not think it makes sense to set BHP control for a well when you all zero historical rates, 

Possibly, but the grammar does not forbid us to do so. 

> I disagree. I think we should not care at all about the control setting when all historical rates are zero, just set STOP always.

I still think it is the control setting matters since that is one entering well equations and causing numerical problems.

Reply:
I think we are talking about two separate but related issues. This PR does not fix, or attempt to fix, the general case of singular constraint (such well running on ORAT with 0 target, or ORAT constraint when the well is in the water zone as discussed in OPM/opm-simulators#4295). This PR only fixes the particular limited case of all-zero historical rates, which seems to be quite common, probably due to the tools used to build the cases.

I still think the current PR makes such cases STOP correctly. As you point out there are more cases that should be STOPped, such as whenever we have ORAT 0.0 for example, but they must be dealt with elsewhere. It cannot be handled in the code changed here, which only sees the WCONHIST keyword. Maybe the simulator code is the correct place, or we could do a post-processing of the Schedule. But that I think is best done in a separate PR.

Reply:
Partly agree, but the new PR will/might cover this PR if I understand correctly, and also this PR might potentially introduce wrong logic regarding WCONHIST specified with BHP constraint, although that can be a rare situation. 

Let us go back to purpose of this PR.  
```
WCONHIST
     'C-4H'      'OPEN'      'ORAT'      0.000      0.000      0.000  5* /
```
What is wrong with this setup if we do not STOP this well?  or understand what circumstances this PR will cause problem?

Reply:
> > But in theory, if any rate constraint is zero, the well should be `STOP`. Is it right?
> 
> I do not think that is necessarily true, if there is no water present in a region, a zero water rate constraint could make sense. In any case, the change here only deals with WCONHIST, so the data looked at are historical data, not constraints. It is common to see one of them being zero. For example from Norne:
> 
> ```
>      'D-1H'      'OPEN'      'RESV'   4347.700      0.000     482594.703  5* /
> ```
> 
> > And furthermore, if WCONHIST is specified with `BHP` control, then the three rates are pure historic values, then well should be able to run.
> 
> While possible, I do not think it makes sense to set BHP control for a well when you all zero historical rates, but if you really want to, you can do it with WCONPROD instead. I think it is better to set STOP for that well when using WCONHIST in that case (as done here).
> 
> > My current feeling is that we can only process to `STOP` wells with zero rate constraint only when we are sure what actual rate constraints are used (e.g. WHISCTL can change the control for all the historic producers), possibly after parsing the whole `SCHEDULE` section.
> 
> I disagree. I think we should not care at all about the control setting when all historical rates are zero, just set STOP always.
> 
> @vkip @tskille do you have an opinion?

My opinion: Check the control and leave OPEN when possible (in particular, 'RESV 0 0 0' and 'BHP 0 0 0' should be fine).

Reply:
To handle the "BHP 0 0 0 stays open" case is not really possible at the level of the method I have modified here, since WHISTCTL must be accounted for. So I will close this in favour of https://github.com/OPM/opm-simulators/pull/4518

Reply:
Thanks for the input, @vkip . The new approach is trying to fix from the simulation side. (https://github.com/OPM/opm-simulators/pull/4518)